src/main/java/org/broadinstitute/dsm/jobs/TestBostonUPSTrackingJob.java

```diff
@@ -58,36 +59,71 @@ public void execute(JobExecutionContext context) throws JobExecutionException {
-                            logger.info("checking tracking status for " + kits.size() + " tracking numbers");
-                            for (DdpKit kit : kits.values()) {
-                                try {
-                                    updateKitStatus(kit, false);
-                                } catch (Exception e) {
-                                    logger.error("Could not update outbound status for " + kit.getExternalOrderNumber(), e);
+            if (ddpInstance != null && ddpInstance.isHasRole()) {
+                logger.info("tracking ups ids for " + ddpInstance.getName());
+                SimpleResult result = inTransaction((conn) -> {
+                    SimpleResult dbVals = new SimpleResult();
+                    String query = SQL_SELECT_KITS + SQL_AVOID_DELIVERD;
+                    try (PreparedStatement stmt = conn.prepareStatement(query)) {
+                        stmt.setString(1, ddpInstance.getDdpInstanceId());
+                        try (ResultSet rs = stmt.executeQuery()) {
+                            while (rs.next()) {
+                                DdpKit kit = new DdpKit(
+                                        rs.getString(DBConstants.DSM_KIT_REQUEST_ID),
+                                        rs.getString("kit." + DBConstants.KIT_LABEL),
+                                        rs.getString("kit." + DBConstants.DSM_TRACKING_TO),
+                                        rs.getString("kit." + DBConstants.TRACKING_RETURN_ID),
+                                        rs.getString("kit." + DBConstants.ERROR),
+                                        rs.getString("kit." + DBConstants.MESSAGE),
+                                        rs.getString("kit." + DBConstants.DSM_RECEIVE_DATE),
+                                        rs.getString("kit." + DBConstants.UPS_TRACKING_STATUS),
+                                        rs.getString("kit." + DBConstants.UPS_TRACKING_DATE),
+                                        rs.getString("kit." + DBConstants.UPS_RETURN_STATUS),
+                                        rs.getString("kit." + DBConstants.UPS_RETURN_DATE),
+                                        rs.getString("req." + DBConstants.COLLABORATOR_PARTICIPANT_ID),
+                                        rs.getString("req." + DBConstants.EXTERNAL_ORDER_NUMBER),
+                                        rs.getBoolean("kit." + DBConstants.CE_ORDER)
+                                );
+                                String type;
+                                if (StringUtils.isNotBlank(kit.getTrackingToId())) {
+                                    type = kit.getUpsTrackingStatus();
+                                    if (StringUtils.isNotBlank(type)) {// get only type from it
+                                        type = type.substring(0, type.indexOf(' '));
+                                    }
+                                    if (!"D".equals(type)) {//don't include delivered ones
```

Comment:
Please use `UPSStatus.DELIVERED_TYPE` instead of `"D"` so that we can find dependencies more easily.  Better still might be a static method on `UPSStatus` like `isDelivered(String statusType)`.

Reply:
Actually even better would be to also add `isDelivered()` to `DdpKit`.  It can use `UPSStatus.isDelivered` internally, and then the information about what you need is right there.  Check out Feature Envy [here](https://blog.codinghorror.com/code-smells/) and [here](https://softwareengineering.stackexchange.com/questions/212128/what-is-a-feature-envy-code-and-why-is-it-considered-a-code-smell).